core/pkg/ingress/types.go

```diff
@@ -0,0 +1,217 @@
+	"k8s.io/ingress/core/pkg/ingress/annotations/ipwhitelist"
+	"k8s.io/ingress/core/pkg/ingress/annotations/proxy"
+	"k8s.io/ingress/core/pkg/ingress/annotations/ratelimit"
+	"k8s.io/ingress/core/pkg/ingress/annotations/rewrite"
+	"k8s.io/ingress/core/pkg/ingress/defaults"
+)
+
+var (
+	// DefaultSSLDirectory defines the location where the SSL certificates will be generated
+	DefaultSSLDirectory = "/ingress-controller/ssl"
+)
+
+// Controller holds the methods to handle an Ingress backend
+type Controller interface {
+	// Start executes a command to start the backend.
+	// The command must run in foreground.
+	Start()
+	// Stop stops the backend
+	Stop() error
+	// Restart signal the backend with a new configuration file returning
+	// the combined output of Stdout and Stderr
+	Restart(data []byte) ([]byte, error)
+	// Tests returns the commands to execute that verifies if the configuration file is valid
+	// Example: nginx -t -c <file>
+	Test(file string) *exec.Cmd
+	// OnUpdate callback invoked from the sync queue https://k8s.io/ingress/core/blob/master/pkg/ingress/controller/controller.go#L387
+	// when an update occurs. This is executed frequently because Ingress
+	// controllers watches changes in:
+	// - Ingresses: main work
+	// - Secrets: referenced from Ingress rules with TLS configured
+	// - ConfigMaps: where the controller reads custom configuration
+	// - Services: referenced from Ingress rules and required to obtain
+	//	 information about ports and annotations
+	// - Endpoints: referenced from Services and what the backend uses
+	//	 to route traffic
+	//
+	// ConfigMap content of --configmap
+	// Configuration returns the translation from Ingress rules containing
+	// information about all the upstreams (service endpoints ) "virtual"
+	// servers (FQDN) and all the locations inside each server. Each
+	// location contains information about all the annotations were configured
+	// https://k8s.io/ingress/core/blob/master/pkg/ingress/types.go#L83
+	OnUpdate(*api.ConfigMap, Configuration) ([]byte, error)
+	// UpstreamDefaults returns the minimum settings required to configure the
+	// communication to upstream servers (endpoints)
+	UpstreamDefaults() defaults.Backend
+	// IsReloadRequired checks if the backend must be reloaded or not.
+	// The parameter contains the new rendered template
+	IsReloadRequired([]byte) bool
+	// Info returns information about the ingress controller
+	// This can include build version, repository, etc.
+	Info() string
+}
+
+// Configuration contains the pieces
+type Configuration struct {
+	HealthzURL           string
+	Upstreams            []*Upstream
+	Servers              []*Server
+	TCPUpstreams         []*Location
+	UDPUpstreams         []*Location
+	PassthroughUpstreams []*SSLPassthroughUpstreams
+}
+
+// Upstream describes one or more remote server/s (endpoints) associated with a service
+type Upstream struct {
+	// Secure indicates if the communication with the endpoint must use TLS
+	Secure bool
```

Comment:
This indicates the protocol to be used between nginx and the endpoint, not from the user.
I will clarify this